File: Semester3/ProgammingParadigms/C4/OCaml/main.ml
(* 1

a)

Mamy tutaj do czynienia z 
a) val x: ‘a -> ‘b -> c’

val f1: (‘a -> ’b -> ’c) -> ‘a -> ‘b -> ‘c

b)
val x: ‘a
val y: ‘a list
val f2: ‘a -> ‘a list -> ‘b -> ‘a list


*)

(*2*)


let rec f x = f x;;

(*3*)

type 'a bt = Empty | Node of 'a * 'a bt * 'a bt;;
let breadthBT tree =
	let rec helper = function
		[] -> []
		| Empty :: tail -> helper tail
		| Node(value, leftSubtree, rightSubtree) :: tail ->	 value :: helper (leftSubtree :: rightSubtree :: tail)
	in helper [tree];;

File: Semester3/ProgammingParadigms/C4/Scala/main.scala
object app extends App {

    sealed trait BT[+A]
    case object Empty extends BT[Nothing]
    case class Node[+A](elem:A, left:BT[A], right:BT[A]) extends BT[A]


    def breadthBT[A](tree: BT[A]) = {
	def helper[A](nodeQueue: List[BT[A]]): List[A] = nodeQueue match {
		case Nil => Nil
		case Empty :: tail => helper(tail)
		case Node(value, leftSubtree, rightSubtree) :: tail => value :: helper(tail ++ List(leftSubtree, rightSubtree))
	}
	helper (List(tree))
}



}